parlai/agents/seq2seq/modules.py

```diff
@@ -278,13 +281,39 @@ def forward(self, xs, ys=None, cand_params=None, prev_enc=None,
         return scores, cand_scores, encoder_states
 
 
+class UnknownDropout(nn.Module):
+    """With set frequency, replaces tokens with unknown token.
+
+    This layer can be used right before an embedding layer to make the model
+    more robust to unknown words at test time.
+    """
+
+    def __init__(self, unknown_idx, probability):
+        """Initialize layer.
+
+        :param unknown_idx: index of unknown token, replace tokens with this
+        :param probability: during training, replaces tokens with unknown token
+                            at this rate.
+        """
+        super().__init__()
+        self.unknown_idx = unknown_idx
+        self.prob = probability
+
+    def forward(self, input):
+        """If training and dropout rate > 0, masks input with unknown token."""
+        if self.training and self.prob > 0:
+            mask = input.new(input.size()).float().uniform_(0, 1) < self.prob
+            input.masked_fill_(mask, self.unknown_idx)
+        return input
+
+
 class RNNEncoder(nn.Module):
     """RNN Encoder."""
 
     def __init__(self, num_features, embeddingsize, hiddensize,
                  padding_idx=0, rnn_class='lstm', numlayers=2, dropout=0.1,
                  bidirectional=False, shared_lt=None, shared_rnn=None,
-                 sparse=False):
+                 input_dropout=0, unknown_idx=4, sparse=False):
```

Comment:
This magic `4` default seems really dangerous. Given that it depends on the actual dict, I feel like it would be safer to have a `None` default, and then throw a `RuntimeException` if `input_dropout != 0`. People who hope to just reuse this module without the `input_dropout` don't have to worry about it, but then people who DO want that feature are explicitly forced to correctly identify their `UNK` token.

Reply:
Or at least leave a comment explaining why it's 4.

Reply:
yeah good call